src/detect-icmpv6-mtu.c

```diff
@@ -0,0 +1,190 @@
+ */
+static int DetectICMPv6mtuMatch (DetectEngineThreadCtx *det_ctx, Packet *p,
+        const Signature *s, const SigMatchCtx *ctx)
+{
+    if (!(PKT_IS_ICMPV6(p)) || PKT_IS_PSEUDOPKT(p))
+        return 0;
+
+    if (!(ICMPV6_HAS_MTU(p)))
+        return 0;
+
+    uint32_t picmpv6mtu = ICMPV6_GET_MTU(p);
+
+    const DetectU32Data *du32 = (const DetectU32Data *)ctx;
+    return DetectU32Match(picmpv6mtu, du32);
+}
+
+/**
+ * \brief this function is used to attach the parsed icmpv6.mtu data into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param icmpv6mtustr pointer to the user provided icmpv6.mtu options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectICMPv6mtuSetup (DetectEngineCtx *de_ctx, Signature *s, const char *icmpv6mtustr)
+{
+    DetectU32Data *icmpv6mtud = DetectU32Parse(icmpv6mtustr);
+    if (icmpv6mtud == NULL)
+        return -1;
+
+    SigMatch *sm = SigMatchAlloc();
+    if (sm == NULL) {
+        DetectICMPv6mtuFree(icmpv6mtud);
+        return -1;
+    }
+
+    sm->type = DETECT_ICMPV6MTU;
+    sm->ctx = (SigMatchCtx *)icmpv6mtud;
+
+    SigMatchAppendSMToList(s, sm, DETECT_SM_LIST_MATCH);
+    s->flags |= SIG_FLAG_REQUIRE_PACKET;
+    s->proto.flags |= DETECT_PROTO_IPV6;
+
+    return 0;
+}
+
+/**
+ * \brief this function will free memory associated with DetectU32Data
+ *
+ * \param ptr pointer to DetectU32Data
+ */
+void DetectICMPv6mtuFree(void *ptr)
+{
+    //DetectU32Data *icmpv6mtud = (DetectU32Data *)ptr;
+    SCFree(ptr);
+}
+
+/* prefilter code */
+
+static void
+PrefilterPacketIcmpv6mtuMatch(DetectEngineThreadCtx *det_ctx, Packet *p, const void *pectx)
+{
+    if (!(PKT_IS_ICMPV6(p)) || PKT_IS_PSEUDOPKT(p))
+        return ;
+
+    if (!(ICMPV6_HAS_MTU(p)))
+        return;
+
+    uint32_t picmpv6mtu = ICMPV6_GET_MTU(p);
+
+    /* during setup Suricata will automatically see if there is another
+     * check that can be added: alproto, sport or dport */
+    const PrefilterPacketHeaderCtx *ctx = pectx;
+    if (PrefilterPacketHeaderExtraMatch(ctx, p) == FALSE)
+        return;
+
+    /* if we match, add all the sigs that use this prefilter. This means
+     * that these will be inspected further */
+    DetectU32Data du32;
+    du32.mode = ctx->v1.u8[0];
+    du32.arg1 = ctx->v1.u32[1];
+    du32.arg2 = UINT32_MAX;
+    if (DetectU32Match(picmpv6mtu, &du32))
+    {
+        SCLogDebug("packet matches icmpv6.mtu/hl %u", picmpv6mtu);
+        PrefilterAddSids(&det_ctx->pmq, ctx->sigs_array, ctx->sigs_cnt);
+    }
+}
+
+static int PrefilterSetupIcmpv6mtu(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
+{
+    return PrefilterSetupPacketHeader(de_ctx, sgh, DETECT_ICMPV6MTU,
+            PrefilterPacketU32Set,
+            PrefilterPacketU32Compare,
+            PrefilterPacketIcmpv6mtuMatch);
+}
+
+static bool PrefilterIcmpv6mtuIsPrefilterable(const Signature *s)
+{
+    const SigMatch *sm;
```

Comment:
This could call a generic function `PrefilteIsPrefilterableByType(DETECT_ICMPV6MTU)`

What is preferred ?

Reply:
ok